lbrynet/blob/blob_file.py

```diff
@@ -16,19 +16,21 @@
 blobhash_length = get_lbry_hash_obj().digest_size * 2
 
 
-def is_valid_hashcharacter(char):
-    return char in "0123456789abcdef"
-
+def is_hex_character(char):
+    try:
+        return 16 > int(char, 16) >= 0
+    except ValueError:  # Raised if non base16 string given
+        return False
```

Comment:
This make the code about 7x slower:
```
Python 3.7.0 (default, Aug  6 2018, 20:07:46) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import timeit
>>> timeit.timeit('"f" in "0123456789abcdef"', number=10000)
0.0009198940824717283
>>> newway = """
... try:
...   16 > int('f', 16) >= 0
... except ValueError:
...   pass
... """
>>> timeit.timeit(newway, number=10000)
0.007640283089131117
```

Reply:
That's weird. I *know* the fastest way would be to simply to get the single `char` and actually have it index a base16 array (or check to see if it indexes properly without the array), but since that's not available in this case, I guess converting it causes more overhead than I thought.

Reply:
Both the `in` operation and the `int('f', 16)` operations are probably very fast, the difference is that the `in` is at the language / assembly level while `int()` requires calling a function:
```
$ python3.7
Python 3.7.0 (default, Aug  6 2018, 20:07:46) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from dis import dis
>>> dis('"f" in "0123456789abcdef"')
  1           0 LOAD_CONST               0 ('f')
              2 LOAD_CONST               1 ('0123456789abcdef')
              4 COMPARE_OP               6 (in)
              6 RETURN_VALUE
>>> dis("16 > int('f', 16) >= 0")
  1           0 LOAD_CONST               0 (16)
              2 LOAD_NAME                0 (int)
              4 LOAD_CONST               1 ('f')
              6 LOAD_CONST               0 (16)
              8 CALL_FUNCTION            2
             10 DUP_TOP
             12 ROT_THREE
             14 COMPARE_OP               4 (>)
             16 JUMP_IF_FALSE_OR_POP    24
             18 LOAD_CONST               2 (0)
             20 COMPARE_OP               5 (>=)
             22 RETURN_VALUE
        >>   24 ROT_TWO
             26 POP_TOP
             28 RETURN_VALUE
```
You can see above how much more work is the `int()` approach.

Reply:
What's a "Python 3 STL"? Do you mean AST (Abstract Syntax Tree)?

`int` is not a reserved word and not part of any AST, it's a function like any other function and just happens to be built-in.

You can even treat it like a regular variable and override it:
```
$ python3.7
Python 3.7.0 (default, Aug  6 2018, 20:07:46) 
[GCC 5.4.0 20160609] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> int = 33
>>> int
33
```